pkg/backup/backup_pv_action.go

```diff
@@ -0,0 +1,76 @@
+/*
+Copyright 2017 Heptio Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package backup
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"github.com/heptio/ark/pkg/util/collections"
+)
+
+// backupPVAction inspects a PersistentVolumeClaim for the PersistentVolume
+// that it references and backs it up
+type backupPVAction struct {
+}
+
+var _ Action = &backupPVAction{}
+
+func NewBackupPVAction() Action {
+	return &backupPVAction{}
+}
+
+// Execute finds the PersistentVolume referenced by the provided
+// PersistentVolumeClaim and backs it up
+func (a *backupPVAction) Execute(ctx *backupContext, pvc map[string]interface{}, backupper itemBackupper) error {
+	ctx.log("running GetPVAction")
+
+	volumeName, err := collections.GetString(pvc, "spec.volumeName")
+	if err != nil {
+		return err
+	}
+
+	// resolve the GVR (need to use kb.discoveryHelper)
+	ctx.log("getting GVR for PersistentVolume %s\n", volumeName)
+	gvr, resource, err := ctx.discoveryHelper.ResourceFor(schema.GroupVersionResource{Resource: "persistentvolumes"})
+	if err != nil {
+		ctx.log("error getting GVR: %s\n", err)
+	}
+	gr := gvr.GroupResource()
+
+	// get a dynamic client
+	ctx.log("getting client for GVR & Resource %s & %s\n", gvr.String(), resource.String())
+	client, err := ctx.dynamicFactory.ClientForGroupVersionResource(gvr, resource, "")
+	if err != nil {
+		ctx.log("error getting client: %s\n", err)
+	}
+
+	// get the item
+	ctx.log("getting PV %s\n", volumeName)
+	pv, err := client.Get(volumeName, metav1.GetOptions{})
+	if err != nil {
+		ctx.log("error getting PV: %s\n", err)
+	}
+
+	// back it up
+	ctx.log("backing up PV %s\n", volumeName)
+	if err := backupper.backupItem(ctx, pv.UnstructuredContent(), gr); err != nil {
+		ctx.log("error backing up PV: %s\n", err)
```

Comment:
Shouldn't we return the error?

Reply:
yes, a bunch of the logging/error handling in here was messed up so I fixed it.